### DISCUSSION TITLE
New playgrounds

### DISCUSSION THREAD

**Discussion by nalgeon** [codapi-org/support#1] [Ideas]
If you'd like to see a specific programming language, database system, CLI tool or other software available as a Codapi playground — leave a comment!

**Comment by Delta456** [comment]
Hi Anton! :wave:  Spectacular work on `codeapi` :tada: 

I am one of the developers of [V Programming Language](https://vlang.io/). I want V Language to be available as a `codeapi` playground. Please let me know if you need any help to implement it.

**Reply by nalgeon** [reply]
Hi Swastik, thanks for reaching out! I've added the V sandbox. You can try it out on localhost:3000 with the following example:

```html
<pre><code>fn main() { println('Hello, world') }</code></pre>

<codapi-snippet sandbox="v" editor="basic">
</codapi-snippet>
```

**Reply by nalgeon** [reply]
You can also try it here: https://codapi.org/v

**Reply by Delta456** [reply]
> Hi Swastik, thanks for reaching out! I've added the V sandbox. You can try it out on localhost:3000 with the following example:
> 
> ```
> <pre><code>fn main() { println('Hello, world') }</code></pre>
> 
> <codapi-snippet sandbox="v" editor="basic">
> </codapi-snippet>
> ```

Thanks! Is this built updated with every weekly release or every version release?

**Reply by nalgeon** [reply]
Updates are done manually, so definitely not every week :) I'll try to keep up with minor releases.

**Reply by Delta456** [reply]
> Updates are done manually, so definitely not every week :) I'll try to keep up with minor releases.

Ah, I see. Let me know if you need help with updates and more.

**Reply by nalgeon** [reply]
Hey Swastik, I've been thinking about writing a (shortened) interactive V tutorial based on the [docs](https://github.com/vlang/v/blob/master/doc/docs.md). Would you be interested in hosting it on the V site? Or maybe providing a link to the tutorial on [codapi.org/try](https://codapi.org/try/) if you prefer not to host it it yourself?

**Reply by Delta456** [reply]
I will take a look into it soon.

**Comment by joe-chelladurai** [comment]
Anton! Thanks for your work :) I would like to request support for R and for installing R packages for working with data and data visualization. Thanks again!

**Reply by nalgeon** [reply]
Sure! I added an R sandbox with Tidyverse installed: https://codapi.org/r

**Reply by joe-chelladurai** [reply]
The R sandbox works great! Thank you! 

However, I'm unable to use tidyverse or it's bundled packages such as ggplot and dplyr. 



I'm also unable to get plots even without any packages. It seems like the plot is saving as a pdf. I like the example where in the python sandbox we can get an svg instead. Perhaps that would be a better option?

<img width="600" alt="Screenshot 2024-02-21 at 8 01 06 AM" src="https://github.com/codapi-org/support/assets/58997358/046d4459-4127-47d5-9805-4b3910d7c289">

Tidyverse functions like select (part of dplyr package) and ggplot (part of ggplot2 package) are also not working. The dplyr and ggplot2 packages come bundled with tidyverse.

<img width="600" alt="Screenshot 2024-02-21 at 8 02 57 AM" src="https://github.com/codapi-org/support/assets/58997358/4141e227-4c6e-497f-896d-e77eebec36fa">

<img width="600" alt="Screenshot 2024-02-21 at 8 02 30 AM" src="https://github.com/codapi-org/support/assets/58997358/a5b75a47-798e-489b-a07c-dfec5f1979d5">

<img width="600" alt="Screenshot 2024-02-21 at 8 03 19 AM" src="https://github.com/codapi-org/support/assets/58997358/8340313a-6ef4-463c-9bc1-947ad8efbc63">


When I run `sessionInfo()` it shows me the packages installed, but it seems like `tidyverse` isn't showing up.

<img width="600" alt="Screenshot 2024-02-21 at 8 01 06 AM" src="https://github.com/codapi-org/support/assets/58997358/15ccbcbd-3eda-4d22-ad8e-d05225a336b5">


Thanks again for your work!

I'm trying to learn how your project works and I'd love to contribute when I figure things out.

**Reply by nalgeon** [reply]
Honestly, I know nothing about R :) But I think you have to import a package to make it work.

Here is some (apparently) working code:

```r
library(ggplot2)
library(svglite)

# Create your plot
plot <- ggplot(data=mtcars, aes(x = mpg, y = hp))

# Use svgstring() to capture SVG output
svg <- svgstring()
print(plot)

# Print SVG content to stdout
cat(svg())
```

Note that you must set the [`output-mode`](https://github.com/nalgeon/codapi-js/blob/main/docs/output-modes.md) to `svg`, otherwise it will display raw SVG content.

**Reply by joe-chelladurai** [reply]
This worked! Thank you!

Is there a possibility to hide the package loading from the user. It may feel like a little too much for the user to load packages for each code chunk. Or if there could be an option to specify which packages to load by putting it inside the codapi snippet? 

![Screenshot 2024-02-21 at 9 56 47 AM](https://github.com/codapi-org/support/assets/58997358/65f74d53-abda-4f7d-8401-fa2f441e658d)

**Reply by nalgeon** [reply]
Sure! You can use [templates](https://github.com/nalgeon/codapi-js/blob/main/docs/templates.md) to hide the boilerplate code and show only the relevant part.

**Reply by joe-chelladurai** [reply]
This is exactly what I was looking for! Very cool :)

**Reply by joe-chelladurai** [reply]
Hi Anton, I would like to request if codapi could have other popular R packages installed. I would like to request `easystats` `gt` and  `tidymodels`. Thank you :)

**Reply by nalgeon** [reply]
Installed them. I'm not sure if they work well though: these packages seem to be huge and very memory (and cpu) hungry. If you encounter further errors, it would really help if you could provide the full source code of a program you'd like to run.

**Reply by joe-chelladurai** [reply]
They are working well so far and I don't get any errors. Thank you!

**Comment by jsjoeio** [comment]
Hi! What about Git? I would love to be able to advanced Git in the browser and wonder if there's a way to do it using codapi

**Reply by nalgeon** [reply]
Could you please give a specific example of what you would like to do with git?

**Reply by jsjoeio** [reply]
it'd be cool to teach someone how to use [git worktree](https://git-scm.com/docs/git-worktree) in a little playground:

"Run the following command:
```shell
git worktree add ../hotfix
```
This will do X..."

Maybe this use case doesn't fit a codapi playground though 🤔

**Reply by nalgeon** [reply]
I think it's possible. I'll prepare the sandbox and let you know.

**Reply by nalgeon** [reply]
Done! Now you can use `git` in the [shell](https://codapi.org/shell/) sandbox. Note that only the `/tmp` directory is writable.

Here is a simple example:

```sh
cd /tmp && mkdir repo && cd repo

git init > /dev/null 2>&1

echo "hello, world" > message.txt
git add message.txt
git commit -m "add message" > /dev/null

echo "hello, world!" > message.txt
git commit -am "edit message" > /dev/null

git log
```

**Reply by jsjoeio** [reply]
that's incredible!!!!!! 

<img width="852" alt="image" src="https://github.com/codapi-org/support/assets/3806031/a1df80dd-acd9-4dfe-bbc5-aa23fbf0354a">

**Comment by bhavaniravi** [comment]
I'd really love to have JSONata playgrounds. Recently using it extensively for data transformations

**Comment by Bhupesh-V** [comment]
Hey Anton, awesome work with codapi.

Would like to request Elixir. A perfect (public) use-case seems to be for the [elixirbyexample.com](elixirbyexample.com) guide, which seems to be a WIP at the moment (https://github.com/technusm1/elixir-by-example).

Let me know how can I help, new to elixir, but I can definitely find folks who would be willing to help.

**Reply by nalgeon** [reply]
Thank you, Bhupesh! Here you go: https://codapi.org/elixir. As for help — please spread the word among the Elixir developers.

**Comment by piratabrasileiro1** [comment]
HAXE, please

**Comment by ethan-sh-lee** [comment]
Anton! Thanks for your work!!
I would like to request support for kotlin and scala.
Thanks again😁

**Reply by nalgeon** [reply]
Thank you, Ethan! I plan to look into Kotlin later. As for Scala, could you please share your use case for it?

**Reply by ethan-sh-lee** [reply]
> Thank you, Ethan! I plan to look into Kotlin later. As for Scala, could you please share your use case for it?

A usecase is a functional programming playground based on the [Functional Programming in Scala](https://www.manning.com/books/functional-programming-in-scala) and [Functional Programming in Kotlin](https://www.manning.com/books/functional-programming-in-kotlin). 

Thank you!

**Comment by yanecc** [comment]
Thanks for your work. Could you please add support for the [Crystal](https://crystal-lang.org) programming language and [Kroki](https://kroki.io)?

**Comment by ldemailly** [comment]
My language is just getting started (well... 50th+ releases already) but would be neat to have [Grol](https://github.com/grol-io/grol#grol) support, I could use it to build:
- grol-io/grol#56

**Reply by habere-et-dispertire** [reply]
Hello @nalgeon ! 🍫

My use case is for blogging [explorable explanations](https://en.wikipedia.org/wiki/Explorable_explanation) in [raku](https://raku.org/). 🦋

* Is there intergration with blogging platforms such as dev.to?
* How is syntax highlighting handled?

In particular I would love to help bring [instructional material ](https://course.raku.org/) to life and to help authors ensure example code undergoes sanity checking/cursory validation as part of the authoring process.

Thank you for helping transform the learning experience away from a static, non-interactive experience, and for putting [dogfood](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) back on the menu. 🥫🐕

I stumbled upon this project via a GitHub star of @steffan153 -- and I stumbled upon him via [his participation](https://exercism.org/profiles/Steffan153) on Exercism's raku track.

**Reply by nalgeon** [reply]
Hey @habere-et-dispertire, thank you for your interest in Codapi! You [can use it](https://github.com/nalgeon/codapi-js/blob/main/docs/code-links.md) with dev.to, although in a very limited way (dev.to forbids running third-party JavaScript). As for course.raku.org, it does not seem to be handled by dev.to, so you can probably use the [fully functional widget](https://github.com/nalgeon/codapi-js/blob/main/docs/html.md).

**Reply by habere-et-dispertire** [reply]
Thanks for the explanation. 👍
I tried but I guess we need a raku sandbox first? 📦

https://codapi.org/embed/?sandbox=raku&code=data%3A%3Bbase64%2CK06sVFD3SM3JyddRCM8vyklRVLcGAA%3D%3D

~~~~text
say 'Hello, World!';

# 404 - Unknown sandbox or command.
~~~~

In case it helps -- raku docker images [are available](https://github.com/Raku/docker).

**Reply by nalgeon** [reply]
Oh, right :) I've [added it now](https://codapi.org/embed/?sandbox=raku&code=data%3A%3Bbase64%2CdY5BCoMwEEX3OcWHLrLrAax1W1wZRA%2BQQrCgJiVqQVIP4rbFY3gZL%2BAVSqK01NLVZOb9yZtM87LkGozrSmgYAgC6KQSQRAwwCOEX6iYC%2BAWXWYDORWqVCwkLYECnfqC4L%2B0X5zKDgY%2BY5w2Y0AXipqpxUmBtfVEScXNuYT%2FtCKl4u96xv9rigYaY%2BsFt0wPZIUoTliYe5vHxIfP4XMzHwIKpH9zEqpfJGvojcKv2tF%2FDG20U9rlxrKkX).

**Reply by habere-et-dispertire** [reply]
The sandbox compiler appears [six months old](https://codapi.org/embed/?sandbox=raku&code=data%3A%3Bbase64%2CdY5BCoMwEEX3OcWHLrLrAax1W1wZRA%2BQQrCgJiVqQVIP4rbFY3gZL%2BAVSqK01NLVZOb9yZtM87LkGozrSmgYAgC6KQSQRAwwCOEX6iYC%2BAWXWYDORWqVCwkLYECnfqC4L%2B0X5zKDgY%2BY5w2Y0AXipqpxUmBtfVEScXNuYT%2FtCKl4u96xv9rigYaY%2BsFt0wPZIUoTliYe5vHxIfP4XMzHwIKpH9zEqpfJGvojcKv2tF%2FDG20U9rlxrKkX).

~~~~text
say $*RAKU.compiler
rakudo (2024.04)
~~~~

Any chance we can get it to current 2024.10 ?
No pressure. 🏋️‍♂️
Happy to help -- just send a pointer. ⛑ 😀

**Reply by nalgeon** [reply]
Updated to 2024.09, which is as recent as possible for now, given the way the sandbox is implemented.

**Comment by nalgeon** [comment]
I think it's better to have [separate discussions](https://github.com/orgs/codapi-org/discussions/new?category=ideas&title=New+playground+-+[...]&body=I+will+use+it+for+[...]) for new playgrounds, so I'm going to close this one.
